frontend/src/core/codemirror/rtc/extension.ts

```diff
@@ -12,126 +17,167 @@ import {
+awareness.addListener((updates, origin) => {
+  const changes = [...updates.added, ...updates.removed, ...updates.updated];
+  if (origin === "local") {
+    log("awareness changes", changes);
+    const ws = getWs();
+    ws.send(awareness.encode(changes));
+  }
+});
 
 export function realTimeCollaboration(
   cellId: CellId,
-  updateCellCode: (code: string) => void,
+  _updateCellCode: (code: string) => void,
   initialCode = "",
-): { extension: Extension; code: string } {
+): { extension: Extension; code?: string } {
   if (isWasm()) {
     return {
       extension: [],
       code: initialCode,
     };
   }
 
-  const manager = CellProviderManager.getInstance();
-  const { ytext, ylanguage, provider } = manager.getOrCreateProvider(
-    cellId,
-    initialCode,
-  );
-
-  // Code sync plugin
-  const codeSync = ViewPlugin.define((view) => {
-    const handleSync = (isSynced: boolean) => {
-      Logger.debug(`RTC sync=${isSynced}, ytext.length=${ytext.toJSON()}`);
-
-      // If it's not synced, update the editor code
-      if (!isSynced) {
-        const code = ytext.toJSON();
-        view.dispatch({
-          changes: {
-            from: 0,
-            to: view.state.doc.length,
-            insert: code,
-          },
-        });
-      }
+  // Connect if not already connected
+  const ws = getWs();
+  if (ws.shouldReconnect) {
+    log("connecting to websocket");
+    ws.reconnect();
+  }
 
-      // If it's synced, update the cell code
-      if (isSynced) {
-        updateCellCode(ytext.toJSON());
-
-        // If the cell is new, insert the initial code
-        // Only do this once when we first connect and the document is empty
-        if (initialCode && ytext.length === 0) {
-          // Set a flag to prevent multiple insertions
-          const hasInserted = ytext.doc
-            ?.getMap("_metadata")
-            .get("initialCodeInserted");
-          if (!hasInserted) {
-            ytext.doc?.transact(() => {
-              ytext.delete(0, ytext.length);
-              ytext.insert(0, initialCode);
-              // Mark that we've inserted the initial code
-              ytext.doc?.getMap("_metadata").set("initialCodeInserted", true);
-            });
-          }
-        }
-      }
-    };
+  return {
+    extension: [
+      languageObserverExtension(),
+      languageListenerExtension(),
+      LoroSyncPlugin(doc, ["codes", cellId], () =>
+        doc.getMap("codes").getOrCreateContainer(cellId, new LoroText()),
+      ),
+    ],
+  };
+}
 
-    // Wait for provider to connect before initializing with local code
-    provider.on("sync", handleSync);
-    return {
-      destroy() {
-        provider.off("sync", handleSync);
-      },
-    };
-  });
+/**
+ * Create a view plugin to observe language changes
+ * When the language changes, we need to update our local language
+ * Server -> Local
+ * @param cellId - The cell id
+ * @returns Extension
+ */
+export function languageObserverExtension() {
```

Comment:
Wondering if per language specification adds unneeded complexity.

Consider someone editing from vim connected tot he WS endpoint, they'll be editing in python, but have to know the correct document to set

Reply:
*vscode over vim